src/Microsoft.AspNetCore.Mvc.Testing/WebApplicationFactory.cs

```diff
@@ -0,0 +1,344 @@
+            {
+                _configuration(builder);
+                configuration(builder);
+            },
+            new WebApplicationFactoryClientOptions(ClientOptions));
+
+            _derivedFactories.Add(factory);
+
+            return factory;
+        }
+
+        private void EnsureServer()
+        {
+            if (_server != null)
+            {
+                return;
+            }
+
+            EnsureDepsFile();
+
+
+            var builder = CreateWebHostBuilder();
+            SetContentRoot(builder);
+            ConfigureWebHost(builder);
+            _configuration(builder);
+            _server = CreateServer(builder);
+        }
+
+        private void SetContentRoot(IWebHostBuilder builder)
+        {
+            var metadataAttributes = GetContentRootMetadataAttributes(
+                typeof(TEntryPoint).Assembly.FullName,
+                typeof(TEntryPoint).Assembly.GetName().Name);
+
+            string contentRoot = null;
+            for (var i = 0; i < metadataAttributes.Length; i++)
+            {
+                var contentRootAttribute = metadataAttributes[i];
+                var contentRootCandidate = Path.Combine(
+                    AppContext.BaseDirectory,
+                    contentRootAttribute.ContentRootPath);
+
+                var contentRootMarker = Path.Combine(
+                    contentRootCandidate,
+                    Path.GetFileName(contentRootAttribute.ContentRootTest));
+
+                if (File.Exists(contentRootMarker))
+                {
+                    contentRoot = contentRootCandidate;
+                    break;
+                }
+            }
+
+            if (contentRoot != null)
+            {
+                builder.UseContentRoot(contentRoot);
+            }
+            else
+            {
+                builder.UseSolutionRelativeContentRoot(typeof(TEntryPoint).Assembly.GetName().Name);
+            }
+        }
+
+        private WebApplicationFactoryContentRootAttribute[] GetContentRootMetadataAttributes(
+            string tEntryPointAssemblyFullName,
+            string tEntryPointAssemblyName)
+        {
+            var testAssembly = GetTestAssemblies();
+            var metadataAttributes = testAssembly
+                .SelectMany(a => a.GetCustomAttributes<WebApplicationFactoryContentRootAttribute>())
+                .Where(a => string.Equals(a.Key, tEntryPointAssemblyFullName, StringComparison.OrdinalIgnoreCase) ||
+                            string.Equals(a.Key, tEntryPointAssemblyName, StringComparison.OrdinalIgnoreCase))
+                .OrderBy(a => a.Priority)
+                .ToArray();
+
+            return metadataAttributes;
+        }
+
+        /// <summary>
+        /// Gets the assemblies containing the functional tests where the
+        /// <see cref="WebApplicationFactoryContentRootAttribute"/> define
+        /// content root to use for the given <typeparamref name="TEntryPoint"/>.
```

Comment:
This doc comment seems to have grammatical errors. However, I'm not sure exactly what it's trying to say, so I can't offer suggestions.